keydata.go

```diff
@@ -19,10 +19,506 @@
 )
+
+// KeyCreationData is the data required to create a new KeyData object.
+// It should be produced by a platform implementation.
+type KeyCreationData struct {
+	PlatformKeyData
+	PlatformName string // Name of the platform that produced this data
+
+	// AuxiliaryKey is a key used to authorize changes to the key data.
+	// It must match the key protected inside PlatformKeyData.EncryptedPayload.
+	AuxiliaryKey AuxiliaryKey
+
+	// SnapModelAuthHash is the digest algorithm used for HMACs of Snap
+	// device models, and also the digest algorithm used to produce the
+	// key digest.
+	SnapModelAuthHash crypto.Hash
+}
+
+// KeyID is the unique ID for a KeyData object. It is used to facilitate the
+// sharing of state between the early boot environment and OS runtime.
+type KeyID struct {
+	Loader string // "file" or "luks2"
+	Device uint64 // ID of block device
+
+	// Name identifies a key on a block device. The interpretation of this
+	// depends on the value of Loader. If Loader is "file" then this is the
+	// path of the file on the block device. If Loader is "luks2" then it
+	// refers to a LUKS2 token by its name (the value of "ubuntu_fde_name").
+	Name string
+
+	Revision uint64
+}
+
+func (x *KeyID) Equals(y *KeyID) bool {
+	return *x == *y
+}
+
+func (id *KeyID) String() string {
+	return fmt.Sprintf("%s:[%d:%d]:%s@%d", id.Loader, unix.Major(id.Device), unix.Minor(id.Device), id.Name, id.Revision)
+}
+
+// KeyDataWriter is an interface used by KeyData to write the data to
+// persistent storage in an atomic way.
+type KeyDataWriter interface {
+	io.Writer
+	Cancel() error
+	Commit() error
+}
+
+// KeyDataReader is an interface used to read and decode a KeyData
+// from persistent storage.
+type KeyDataReader interface {
+	io.Reader
+	ID() *KeyID
+}
+
+type hashAlg struct {
+	crypto.Hash
+}
+
+func (a hashAlg) MarshalJSON() ([]byte, error) {
+	var s string
+
+	switch a.Hash {
+	case crypto.SHA1:
+		s = "sha1"
+	case crypto.SHA224:
+		s = "sha224"
+	case crypto.SHA256:
+		s = "sha256"
+	case crypto.SHA384:
+		s = "sha384"
+	case crypto.SHA512:
+		s = "sha512"
+	default:
+		return nil, errors.New("invalid hash algorithm")
+	}
+
+	return json.Marshal(s)
+}
+
+func (a *hashAlg) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	switch s {
+	case "sha1":
+		a.Hash = crypto.SHA1
+	case "sha224":
+		a.Hash = crypto.SHA224
+	case "sha256":
+		a.Hash = crypto.SHA256
+	case "sha384":
+		a.Hash = crypto.SHA384
+	case "sha512":
+		a.Hash = crypto.SHA512
+	default:
+		a.Hash = crypto.Hash(0)
+	}
+
+	return nil
+}
+
+type snapModelHMAC []byte
+
+type snapModelHMACList []snapModelHMAC
+
+func (l snapModelHMACList) contains(h snapModelHMAC) bool {
+	for _, v := range l {
+		if bytes.Equal(v, h) {
+			return true
+		}
+	}
+	return false
+}
+
+type authorizedSnapModels struct {
+	Alg       hashAlg           `json:"alg"`
+	KeyDigest []byte            `json:"key_digest"`
```

Comment:
Samuele mentioned this in the design document as well. The convention here follows what is in the LUKS2 format specification as that's where the data will eventually end up being stored. The convention there is key_digest rather than key-digest or keyDigest.